File: javax/src/test/java/me/panpf/javax/test/lang/ClassTest.java
/*
 * Copyright (C) 2018 Peng fei Pan <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.panpf.javax.test.lang;

import me.panpf.javax.lang.Classx;
import org.junit.Assert;
import org.junit.Test;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("unused")
public class ClassTest {

    public static class TestField1 {
        public String testFiled11 = "field11";
        public String testFiled12 = "field12";
    }

    public static class TestField2 extends TestField1 {
        public String testFiled21 = "field21";
        public String testFiled22 = "field22";
    }

    public static class TestField3 extends TestField2 {
        public String testFiled31 = "field31";
        public String testFiled32 = "field32";
    }

    public static class TestMethod {

        private String update;

        public String getUpdate() {
            return update;
        }

        public void setUpdate(String update) {
            this.update = update;
        }
    }

    public static class TestConstructor {

        public TestConstructor() {
        }

        public TestConstructor(String param1) {
        }

        public TestConstructor(String param1, int param2) {
        }
    }

    public static class TestType {
        String[] strings = new String[]{};
        List<String> stringList = new ArrayList<String>();
    }

    @Test
    public void testField() {
        TestField1 testClass = new TestField1();

        try {
            Classx.setFieldValue(testClass, "testFiled11", "field11x");
        } catch (NoSuchFieldException e) {
            throw new IllegalArgumentException(e);
        }

        try {
            Assert.assertEquals(Classx.getFieldValue(testClass, "testFiled11"), "field11x");
        } catch (NoSuchFieldException e) {
            throw new IllegalArgumentException(e);
        }

        Field[] fields = Classx.getFieldsWithParent(TestField3.class);
        Assert.assertNotNull(fields);
        Assert.assertEquals(fields.length, 6);

        Field[] field2 = Classx.getFieldsWithParent(TestField3.class, 1);
        Assert.assertNotNull(field2);
        Assert.assertEquals(field2.length, 4);
    }

    @Test
    public void testMethod() {
        TestMethod testMethod = new TestMethod();

        try {
            Classx.callMethod(testMethod, "setUpdate", "testMethod");
        } catch (NoSuchMethodException e) {
            throw new IllegalArgumentException(e);
        }

        try {
            Assert.assertEquals(Classx.callMethod(testMethod, "getUpdate"), "testMethod");
        } catch (NoSuchMethodException e) {
            throw new IllegalArgumentException(e);
        }

        try {
            Classx.callMethod(testMethod, Classx.getMethodWithParent(testMethod.getClass(), "setUpdate", String.class), "testMethod2");
        } catch (NoSuchMethodException e) {
            throw new IllegalArgumentException(e);
        }

        try {
            Assert.assertEquals(Classx.callMethod(testMethod, Classx.getMethodWithParent(testMethod.getClass(), "getUpdate")), "testMethod2");
        } catch (NoSuchMethodException e) {
            throw new IllegalArgumentException(e);
        }

        Method[] methods = Classx.getMethodsWithParent(TestMethod.class);
        Assert.assertNotNull(methods);
        Assert.assertEquals(methods.length, 14);

        Method[] methods2 = Classx.getMethodsWithParent(TestMethod.class, 0);
        Assert.assertNotNull(methods2);
        Assert.assertEquals(methods2.length, 2);
    }

    @Test
    public void testConstructor() {
        try {
            Assert.assertNotNull(Classx.getConstructorWithParent(TestConstructor.class, String.class));
        } catch (NoSuchMethodException e) {
            throw new IllegalArgumentException(e);
        }

        Assert.assertEquals(Classx.getConstructorsWithParent(TestConstructor.class).length, 4);

        Assert.assertEquals(Classx.getConstructorsWithParent(TestConstructor.class, 0).length, 3);
    }

    @Test
    public void testHierarchy() {
        Assert.assertEquals(Classx.getClassHierarchy(TestField3.class).length, 4);
        Assert.assertEquals(Classx.getClassHierarchy(TestField3.class, true).length, 3);
    }

    @Test
    public void testType() {
        try {
            Assert.assertTrue(Classx.isTypeArray(Classx.getFieldWithParent(TestType.class, "strings"), String.class));
        } catch (NoSuchFieldException e) {
            throw new IllegalArgumentException(e);
        }

        try {
            Assert.assertFalse(Classx.isTypeArray(Classx.getFieldWithParent(TestType.class, "strings"), Integer.class));
        } catch (NoSuchFieldException e) {
            throw new IllegalArgumentException(e);
        }

        try {
            Assert.assertTrue(Classx.isTypeCollection(Classx.getFieldWithParent(TestType.class, "stringList"), List.class, String.class));
        } catch (NoSuchFieldException e) {
            throw new IllegalArgumentException(e);
        }

        try {
            Assert.assertFalse(Classx.isTypeCollection(Classx.getFieldWithParent(TestType.class, "stringList"), List.class, Integer.class));
        } catch (NoSuchFieldException e) {
            throw new IllegalArgumentException(e);
        }
    }
}

File: javax-kt/src/test/java/me/panpf/javaxkt/test/lang/ClassTest.kt
/*
 * Copyright (C) 2018 Peng fei Pan <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.panpf.javaxkt.test.lang

import me.panpf.javaxkt.lang.*
import org.junit.Assert
import org.junit.Test

import java.util.ArrayList

class ClassTest {

    open class TestField1 {
        var testFiled11 = "field11"
        var testFiled12 = "field12"
    }

    @Suppress("unused")
    open class TestField2 : TestField1() {
        var testFiled21 = "field21"
        var testFiled22 = "field22"
    }

    @Suppress("unused")
    class TestField3 : TestField2() {
        var testFiled31 = "field31"
        var testFiled32 = "field32"
    }

    @Suppress("unused")
    class TestMethod {
        var update: String? = null
    }

    @Suppress("unused")
    class TestConstructor {
        constructor()

        @Suppress("UNUSED_PARAMETER")
        constructor(param1: String)

        @Suppress("UNUSED_PARAMETER")
        constructor(param1: String, param2: Int)
    }

    @Suppress("unused")
    class TestType {
        internal var strings = arrayOf<String>()
        internal var stringList: List<String> = ArrayList()
    }

    @Test
    fun testField() {
        val testClass = TestField1()

        try {
            testClass.setFieldValue("testFiled11", "field11x")
        } catch (e: NoSuchFieldException) {
            throw IllegalArgumentException(e)
        }

        try {
            Assert.assertEquals(testClass.getFieldValue("testFiled11"), "field11x")
        } catch (e: NoSuchFieldException) {
            throw IllegalArgumentException(e)
        }

        val fields = TestField3().getFieldsWithParent().filter { it.name != "\$jacocoData" }.toTypedArray()
        Assert.assertNotNull(fields)
        Assert.assertEquals(fields.size.toLong(), 6)

        val field2 = TestField3().getFieldsWithParent(1).filter { it.name != "\$jacocoData" }.toTypedArray()
        Assert.assertNotNull(field2)
        Assert.assertEquals(field2.size.toLong(), 4)
    }

    @Test
    fun testMethod() {
        val testMethod = TestMethod()

        try {
            testMethod.callMethod("setUpdate", "testMethod")
        } catch (e: NoSuchMethodException) {
            throw IllegalArgumentException(e)
        }

        try {
            Assert.assertEquals(testMethod.callMethod("getUpdate"), "testMethod")
        } catch (e: NoSuchMethodException) {
            throw IllegalArgumentException(e)
        }

        try {
            testMethod.callMethod(testMethod.getMethodWithParent("setUpdate", String::class.java), "testMethod2")
        } catch (e: NoSuchMethodException) {
            throw IllegalArgumentException(e)
        }

        try {
            Assert.assertEquals(testMethod.callMethod(testMethod.getMethodWithParent("getUpdate")), "testMethod2")
        } catch (e: NoSuchMethodException) {
            throw IllegalArgumentException(e)
        }

        val methods = TestMethod().getMethodsWithParent().filter { it.name != "\$jacocoInit" }.toTypedArray()
        Assert.assertNotNull(methods)
        Assert.assertEquals(methods.size.toLong(), 14)

        val methods2 = TestMethod().getMethodsWithParent(0).filter { it.name != "\$jacocoInit" }.toTypedArray()
        Assert.assertNotNull(methods2)
        Assert.assertEquals(methods2.size.toLong(), 2)
    }

    @Test
    fun testConstructor() {
        try {
            Assert.assertNotNull(TestConstructor().getConstructorWithParent(String::class.java))
        } catch (e: NoSuchMethodException) {
            throw IllegalArgumentException(e)
        }

        Assert.assertEquals(TestConstructor().getConstructorsWithParent().size, 4)

        Assert.assertEquals(TestConstructor().getConstructorsWithParent(0).size, 3)
    }

    @Test
    fun testHierarchy() {
        Assert.assertEquals(TestField3().getClassHierarchy().size, 4)
        Assert.assertEquals(TestField3().getClassHierarchy(true).size, 3)
    }

    @Test
    fun testType() {
        try {
            Assert.assertTrue(TestType().getFieldWithParent("strings").isTypeArray(String::class.java))
        } catch (e: NoSuchFieldException) {
            throw IllegalArgumentException(e)
        }

        try {
            Assert.assertFalse(TestType().getFieldWithParent("strings").isTypeArray(Int::class.java))
        } catch (e: NoSuchFieldException) {
            throw IllegalArgumentException(e)
        }

        try {
            Assert.assertTrue(TestType().getFieldWithParent("stringList").isTypeCollection(List::class.java, String::class.java))
        } catch (e: NoSuchFieldException) {
            throw IllegalArgumentException(e)
        }

        try {
            Assert.assertFalse(TestType().getFieldWithParent("stringList").isTypeCollection(List::class.java, Int::class.java))
        } catch (e: NoSuchFieldException) {
            throw IllegalArgumentException(e)
        }
    }
}
